internal/controllers/extension_controller.go

```diff
@@ -0,0 +1,203 @@
+func (*ExtensionReconciler) checkForUnexpectedFieldChange(a, b ocv1alpha1.Extension) bool {
+	a.Status, b.Status = ocv1alpha1.ExtensionStatus{}, ocv1alpha1.ExtensionStatus{}
+	a.Finalizers, b.Finalizers = []string{}, []string{}
+	return !equality.Semantic.DeepEqual(a, b)
+}
+
+// Helper function to do the actual reconcile
+//
+// Today we always return ctrl.Result{} and an error.
+// But in the future we might update this function
+// to return different results (e.g. requeue).
+//
+//nolint:unparam
+func (r *ExtensionReconciler) reconcile(ctx context.Context, ext *ocv1alpha1.Extension) (ctrl.Result, error) {
+	// validate spec
+	if err := validators.ValidateExtensionSpec(ext); err != nil {
+		// Set the TypeInstalled condition to Unknown to indicate that the resolution
+		// hasn't been attempted yet, due to the spec being invalid.
+		ext.Status.InstalledBundleResource = ""
+		setInstalledStatusConditionUnknown(&ext.Status.Conditions, "installation has not been attempted as spec is invalid", ext.GetGeneration())
+		// Set the TypeResolved condition to Unknown to indicate that the resolution
+		// hasn't been attempted yet, due to the spec being invalid.
+		ext.Status.ResolvedBundleResource = ""
+		setResolvedStatusConditionUnknown(&ext.Status.Conditions, "validation has not been attempted as spec is invalid", ext.GetGeneration())
+
+		setDeprecationStatusesUnknown(&ext.Status.Conditions, "deprecation checks have not been attempted as spec is invalid", ext.GetGeneration())
+		return ctrl.Result{}, nil
+	}
+
+	// gather vars for resolution
+	vars, err := r.variables(ctx)
+	if err != nil {
+		ext.Status.InstalledBundleResource = ""
+		setInstalledStatusConditionUnknown(&ext.Status.Conditions, "installation has not been attempted due to failure to gather data for resolution", ext.GetGeneration())
+		ext.Status.ResolvedBundleResource = ""
+		setResolvedStatusConditionFailed(&ext.Status.Conditions, err.Error(), ext.GetGeneration())
+
+		setDeprecationStatusesUnknown(&ext.Status.Conditions, "deprecation checks have not been attempted due to failure to gather data for resolution", ext.GetGeneration())
+		return ctrl.Result{}, err
+	}
+
+	// run resolution
+	_, err = r.Resolver.Solve(vars)
+	if err != nil {
+		ext.Status.InstalledBundleResource = ""
+		setInstalledStatusConditionUnknown(&ext.Status.Conditions, "installation has not been attempted as resolution failed", ext.GetGeneration())
+		ext.Status.ResolvedBundleResource = ""
+		setResolvedStatusConditionFailed(&ext.Status.Conditions, err.Error(), ext.GetGeneration())
+
+		setDeprecationStatusesUnknown(&ext.Status.Conditions, "deprecation checks have not been attempted as resolution failed", ext.GetGeneration())
+		return ctrl.Result{}, err
+	}
+
+	// TODO: kapp-controller integration
+	// * lookup the bundle in the solution/selection that corresponds to the Extension's desired Source
+	// * set the status of the Extension based on the respective deployed application status conditions.
```

Comment:
Is the plan here to still use both rukpak and kapp control and have the paths guarded by feature gates?